test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/resume/resume001.java

```diff
@@ -361,13 +370,17 @@ private int runThis (String argv[], PrintStream out) {
                         break label1;
                 log2("       thread2 is at breakpoint");
 
-
-                log2("      resuming the thread2");
+                log2("       resuming the thread2");
                 thread2.resume();
 
+                log2("       undo the vm.suspend() with vm.resume()");
+                vm.resume();
+            }
+            // These are only needed if we break out of the loop due to an error
+            if (expresult != returnCode0) {
+                vm.resume();
+                vm.resume();  // for case error when both VirtualMachine and the thread2 were suspended
```

Comment:
Pre-existing but I don't understand the comment. Why would you need 2 `vm.resume()` here? If `thread2` was suspended directly don't you need a `thread2.resume()`?

Reply:
First just to clarify a general JDI feature about thread suspending and resuming. You can undo a ThreadReference.suspend() or a thread suspended as the result of an event by dong a vm.resume(). This is documented in the JDI API spec, which talks about suspendCounts and how various APIs and event delivery affect them.

I was  tempted to clean up these vm.resume() calls but opted not to. The point being made in the comment is that worse case thread2 was suspended by a breakpoint or thread2.suspend() and all threads were suspended by a vm.resume() (meaning thread2 could have a suspendCount of 2). Two vm.resumes() take are done to make sure thread2 gets resumed under this situation. However, one of the vm.resume calls could instead be a thread2.resume(). Doing two vm.resume() calls was probably just laziness by the original  authors. It works though.

However, by my accounting at any failure point thread2 never has a suspendCount > 1, so really just one vm.resume() would be enough.

The original code did these two vm.resume() calls unconditionally, but they are not needed if there was no error.

Reply:
The original code had 2 vm.resume() - one on them to match vm.suspend() and 2nd one to allow debugee to continue on error.
Now we have 3 vm.resume() - one is to match vm.suspend() (line 377) and 2 conditional (on error).
Theoretically we can get an error when both vm and thread2 are suspended, so 2 vm.resume() looks reasonable.
Anyway resume() is a nop if the thread is not suspended

Reply:
After reaching the 2nd breakpoint, which suspends thread2, we do a vm.suspend(), which bumps the thread2 suspendCount to 2. However, we do a eventSet.resume() after this, lowering the suspendCount to 1, and there is no error bailout point while the suspendCount is 2. Thus only 1 vm.resume() is needed in the error handling.

Reply:
I think all this discussion about the number of `vm.resume()` calls that are needed
or not needed and the fact that one of those `vm.resume()` calls could be replaced
by a thread2.resume() call _perfectly_ illustrates just how complicated this test is.

Thanks for going thru the effort to get rid of the `sleep()` call. I appreciate it.